### PR TITLE
Patches related to half-double

### PR DESCRIPTION
**Pull request by sqazi**
There is a new payload generator that can be used to generate the tables used in our white paper.  We also added support in hammer tolerance payload generator to evaluate behaviour at higher distances.

These were tested with payload size of 0x100000.  What is the new procedure for adding documentation?
Please review and provide feedback.  Thanks :).

### CODE REVIEW

File: rowhammer_tester/scripts/playbook/payload_generators/half_double_analysis.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        victim_flipped = False
+        victim_errors = 0
+
+        for logical_row in sorted(row_errors_logical.keys()):
+            row, errors = row_errors_logical[logical_row]
+            if logical_row == logical_victim:
+                victim_flipped = True
+                if self.state != HalfDoubleAnalysisState.NOFLIP_DISTANCE_ONE:
+                    if self.state == HalfDoubleAnalysisState.NOFLIP_DISTANCE_TWO:
+                        dilution = math.inf
```
Comment (jedrzejboczar) [inline rowhammer_tester/scripts/playbook/payload_generators/half_double_analysis.py:190]: Need to add `import math` at the beginning as here `math` is undefined.

File: rowhammer_tester/scripts/playbook/payload_generators/half_double_analysis.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            print()
+
+    def summarize(self):
+        print("Bit histogram:")
+        self.print_pair_histogram(self.bit_count)
+        print("Row histogram:")
+        self.print_pair_histogram(self.row_count)
+        print("Victim list:")
+        self.print_pair_histogram(self.victim_list)
+        return
```
Comment (jedrzejboczar) [inline rowhammer_tester/scripts/playbook/payload_generators/half_double_analysis.py:243]: This `return` can be omitted.

**Review by jedrzejboczar — CHANGES_REQUESTED** [review]
Thanks, this looks good, I just added some minor comments. 

We will need to add some documentation of the new functionality before merging, preferably with a link to the white paper ([this link](https://github.com/google/hammer-kit/blob/main/20210525_half_double.pdf) or some other?). To modify the documentation you will need to edit the source files in the RST format. I think it should be enough to update https://github.com/antmicro/litex-rowhammer-tester/blob/master/doc/playbook.rst. The syntax is described e.g. [here](https://sphinx-tutorial.readthedocs.io/step-1/). 

To preview the output you will need to install sphinx and the theme:
```
pip install sphinx 
pip install git+https://github.com/antmicro/sphinx_antmicro_theme
```
Then you should be able to use `python -m sphinx -b html doc build/documentation/html` to generate the HTML files. You can then open `build/documentation/html/index.html` in your browser.

For better experience when modifying documentation I prefer using [sphinx-autobuild](https://pypi.org/project/sphinx-autobuild/). Install it (`pip install sphinx-autobuild`) and run `sphinx-autobuild -b html doc build/documentation/html --re-ignore 'doc/build/.*'` from the main directory. You can then open the ouput in your browser at `http://localhost:8000/` and the documentation will be rebuilt whenever files are modified.

**Comment by jedrzejboczar** [comment]
In https://github.com/antmicro/litex-rowhammer-tester/commit/f8503dcd05b2310d9bc5459c9f551b4d0c66b4a7 I updated instructions on building documentation. You may need to install libcairo2 and Python packages from requirements-doc.txt.

**Comment by sqazi** [comment]
Updated the documentation.

**Comment by jedrzejboczar** [comment]
Thanks, merging.

**Pull request merged**